typings/styled-components.d.ts

```diff
@@ -3,180 +3,205 @@ import { StatelessComponent, ComponentClass } from "react";
 
 type Component<P> = ComponentClass<P> | StatelessComponent<P>;
 
-export interface ThemeProps {
-  theme: any;
+export interface ThemeProps<T> {
+  theme: T;
 }
 
-export type StyledProps<P> = P & ThemeProps;
-export type OuterStyledProps<P> = P & {
-  theme?: Object;
+export type ThemedStyledProps<P, T> = P & ThemeProps<T>;
+export type StyledProps<P> = ThemedStyledProps<P, any>;
+
+export type ThemedOuterStyledProps<P, T> = P & {
+  theme?: T;
   innerRef?: (instance: any) => void;
 };
+export type OuterStyledProps<P> = ThemedOuterStyledProps<P, any>;
 
 export type Interpolation<P> = FlattenInterpolation<P> | ReadonlyArray<FlattenInterpolation<P> | ReadonlyArray<FlattenInterpolation<P>>>;
 export type FlattenInterpolation<P> = InterpolationValue | InterpolationFunction<P>;
 export type InterpolationValue = string | number;
 export type SimpleInterpolation = InterpolationValue | ReadonlyArray<InterpolationValue | ReadonlyArray<InterpolationValue>>;
 export interface InterpolationFunction<P> {
-  (props: StyledProps<P>): Interpolation<P>;
+  (props: P): Interpolation<P>;
```

Comment:
I don't know how this library works at all, but is the idea that an interpolation function can in turn return another interpolation function? Is that a common use case?